Change TestUtil.terminateBackend(...) to wait for terminated process to exit

sehrope opened the pull request:
### All Submissions:

* [X] Have you followed the guidelines in our [Contributing](https://github.com/pgjdbc/pgjdbc/blob/master/CONTRIBUTING.md) document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### New Feature Submissions:

1. [X] Does your submission pass tests?
2. [X] Does `./gradlew autostyleCheck checkstyleAll` pass ?

-------

Adds helpers to wait for a terminated process to exit by repeatedly checking `pg_stat_activity` in a loop for up to five seconds. This needs to be done everywhere that we call `pg_terminate_backend(...)` as it only sends the termination signal without waiting. That makes a race condition for anything that is subsequently testing against the terminated process. Ex: https://github.com/pgjdbc/pgjdbc/pull/1087#issuecomment-798700040

Rather than have to remember to add the `waitForBackendTermination(...)` everywhere that we issue a terminate, I've made it the default and renamed the original method to `terminateBackendNoWait(...)`.

Hopefully this gets rid of some CI gremlins...

sehrope commented:
Force pushed a fix for the "is alive" SQL to include a column as pre-9.4 doesn't allow the column-less SELECT.

vlsi commented on pgjdbc/src/test/java/org/postgresql/replication/ReplicationConnectionTest.java:
```diff
@@ -58,17 +58,12 @@ public void testIsValid() throws Exception {
-    terminateStatement.execute("SELECT pg_terminate_backend(" + backendId + ")");
-    terminateStatement.close();
-    sqlConnection.close();
+    TestUtil.terminateBackend(backendId);
```
Frankly speaking, static assumption on implicit database looks odd.
It would probably be nice to have "database" or "database type" parameter.

For instance, if we add three-way database tests, then the meaning of `terminateBackend(backendId)` becomes moot.

So I don't want to make static `terminateBackend(int)` easily accessible.

WDYT?

sehrope replied:
Yes that'd be nicer longer term. Maybe by changing the signature to take a Connection, unwrapping it as a PGConnection, and then creating a new connection to that host / port with the super user creds. The assumption would be the same super user creds across everything being tested.

I'll take a look if that will work. Right now there's a bunch of overloads and the real implementation already takes a Connection parameter as the connection used to perform the termination action (not the one being terminated) so I had left things as unchanged as possible to avoid changing the signatures too much.

vlsi replied:
Yeah, you are right, adding method to PGConnection would be overkill for this PR. Let's skip it.

sehrope replied:
Ha! Too late I already did :)

Wasn't as bad as I thought and the end result is much cleaner. Rest of your comments re:Duration and try-with-resources are there too.

vlsi replied:
🙀

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/TestUtil.java:
```diff
@@ -980,11 +980,43 @@ public static int getBackendPid(Connection conn) throws SQLException {
+    String sql = "SELECT EXISTS (SELECT * FROM pg_stat_activity WHERE pid = ?) t";
+    try (PreparedStatement stmt = conn.prepareStatement(sql)) {
+      stmt.setInt(1, pid);
+      ResultSet rs = stmt.executeQuery();
+      rs.next();
+      boolean ret = rs.getBoolean(1);
+      rs.close();
+      return ret;
```
try with resources for ResultSet?
```suggestion
      try (ResultSet rs = stmt.executeQuery()) {
        rs.next();
        return rs.getBoolean(1);
      }
```

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/TestUtil.java:
```diff
@@ -980,11 +980,43 @@ public static int getBackendPid(Connection conn) throws SQLException {
+  /**
+   * Wait for a backend process to terminate and return whether it actual terminated within the maximum wait time.
+   */
+  public static boolean waitForBackendTermination(Connection conn, int pid, long maxWaitMillis, long sleepMillis) throws SQLException, InterruptedException {
```
`java.time.Duration` instead of `long`?

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/TestUtil.java:
```diff
@@ -980,11 +980,43 @@ public static int getBackendPid(Connection conn) throws SQLException {
+  }
+
+  public static boolean waitForBackendTermination(Connection conn, int pid) throws SQLException, InterruptedException {
+    // Default to five-seconds max wait time, with 10ms sleeping
+    return waitForBackendTermination(conn, pid, 5 * 1000L, 10);
```
Should the default timeout be like 30 seconds just in case of CI being busy?

sehrope replied:
Sure we can bump that up too. I think I picked 5-seconds because that's what the one place that was actually checking for process termination had.

davecramer commented on pgjdbc/src/test/java/org/postgresql/test/TestUtil.java:
```diff
@@ -971,50 +973,88 @@ public static String queryForString(Connection conn, String sql) throws SQLExcep
+  }
+
+  public static boolean isPidAlive(Connection conn, int pid) throws SQLException {
+    String sql = "SELECT EXISTS (SELECT * FROM pg_stat_activity WHERE pid = ?) t";
```
seems pid is not available on 8.4. We can probably remove this test from 8.4

sehrope replied:
There's `procpid` column that could be used instead. I'll put an if / else on the backend version.

I took a look and the `procpid` was renamed to `pid` in 9.2 so this would have failed for 9.0 and 9.1 as well but we don't have the min our matrix anymore.

sehrope commented:
Rebased and updated to use procpid on <v9.2. Worked against 8.4 locally so I think this should be good.

Assuming it passes through CI okay and no other comments, I'll merge this tomorrow.
